restapi/admin_users.go

```diff
@@ -402,3 +411,109 @@ func getUpdateUserResponse(params admin_api.UpdateUserInfoParams) (*models.User,
 	}
 	return userElem, nil
 }
+
+// addUsersListToGroups iterates over the user list & assigns the requested groups to each user.
+func addUsersListToGroups(ctx context.Context, client MinioAdmin, usersToUpdate []string, groupsToAssign []string) error {
+	parallelGroupsUpdate := func(user, groupToAssign string) chan error {
+		chProcess := make(chan error)
+
+		go func() error {
+			defer close(chProcess)
+
+			userToAddRemove := []string{user}
+
+			updateReturn := updateGroupMembers(ctx, client, groupToAssign, userToAddRemove, false)
+
+			chProcess <- updateReturn
+
+			return updateReturn
+		}()
+
+		return chProcess
+	}
+
+	parallelUserRead := func(user string) chan error {
+		chUserRead := make(chan error)
+
+		go func() {
+			defer close(chUserRead)
+
+			var groupsUpdate []chan error
+
+			for _, groupN := range groupsToAssign {
+				proc := parallelGroupsUpdate(user, groupN)
+				groupsUpdate = append(groupsUpdate, proc)
+			}
+
+			var channelHasError error = nil
+
+			for _, chanRet := range groupsUpdate {
+				locError := <-chanRet
+
+				if locError != nil {
+
+					channelHasError = locError
+				}
+			}
+
+			chUserRead <- channelHasError
```

Comment:
if we have an error we should return right away
Also it would simplify code.
```suggestion
				if chanRet != nil {
				         chUserRead <- chanRet
				         return
				}
			}
```

Reply:
Refactored function